### ISSUE TITLE
Go interface value

### ISSUE THREAD

**Issue by khanhtc1202** [khanhtc1202/til#24]
TL;DR: An interface value that holds a nil concrete value is itself non-nil.

```go
type Explodes interface {
    Bang()
    Boom()
}

// Type Bomb implements Explodes
type Bomb struct {}
func (*Bomb) Bang() {}
func (Bomb) Boom() {}

func main() {
    var bomb *Bomb = nil
    var explodes Explodes = bomb
    println(bomb, explodes) // '0x0 (0x10a7060,0x0)'
    if explodes != nil {
        println("Not nil!") // 'Not nil!' What are we doing here?!?!
        explodes.Bang()     // works fine
        explodes.Boom()     // panic: value method main.Bomb.Boom called using nil *Bomb pointer
    } else {
        println("nil!")     // why don't we end up here?
    }
}
```

What we see: The bomb pointer is 0x0 which is effectively nil, but explodes is the non-nil (0x10a7060,0x0).

Explain: This is because [interface values are fat pointers](https://research.swtch.com/interfaces). The first element of this pair is the pointer to the method dispatch table for the implementation of the `Bomb` interface by the `Explodes` type, and the second element is the address of the actual `Explodes` object, which is `nil`.

The call to `Bang` succeeds because it applies to `pointers to a Bomb`: there is no need to dereference the pointer to call the method. The `Boom` method acts on `a value` and so a call causes pointers to be dereferenced, which causes a panic.

Note that if we had written `var explodes Explodes = nil`, then `!= nil` would have not succeeded. Or we have to nil-check for both the interface value and if non-nil, check the value pointed to by the interface object as below

```go
if explodes != nil && !reflect.ValueOf(explodes).IsNil() {
    println("Not nil!") // we no more end up here
    explodes.Bang()
    explodes.Boom()
} else {
    println("nil!")     // 'nil' -- all good!
}
```

Read more flaws from [here](https://bluxte.net/musings/2018/04/10/go-good-bad-ugly).
